File: tests/algorithms/x/Clojure/maths/binary_exponentiation.clj
(ns main (:refer-clojure :exclude [binary_exp_recursive binary_exp_iterative binary_exp_mod_recursive binary_exp_mod_iterative]))

(require 'clojure.set)

(defn in [x coll]
  (cond (string? coll) (clojure.string/includes? coll x) (map? coll) (contains? coll x) (sequential? coll) (some (fn [e] (= e x)) coll) :else false))

(defn padStart [s w p]
  (loop [out (str s)] (if (< (count out) w) (recur (str p out)) out)))

(defn indexOf [s sub]
  (let [idx (clojure.string/index-of s sub)] (if (nil? idx) -1 idx)))

(defn split [s sep]
  (clojure.string/split s (re-pattern sep)))

(defn toi [s]
  (Integer/parseInt (str s)))

(defn _fetch [url]
  {:data [{:from "" :intensity {:actual 0 :forecast 0 :index ""} :to ""}]})

(def nowSeed (atom (let [s (System/getenv "MOCHI_NOW_SEED")] (if (and s (not (= s ""))) (Integer/parseInt s) 0))))

(declare binary_exp_recursive binary_exp_iterative binary_exp_mod_recursive binary_exp_mod_iterative)

(def ^:dynamic binary_exp_iterative_b nil)

(def ^:dynamic binary_exp_iterative_e nil)

(def ^:dynamic binary_exp_iterative_result nil)

(def ^:dynamic binary_exp_mod_iterative_b nil)

(def ^:dynamic binary_exp_mod_iterative_e nil)

(def ^:dynamic binary_exp_mod_iterative_result nil)

(def ^:dynamic binary_exp_mod_recursive_r nil)

(def ^:dynamic binary_exp_recursive_half nil)

(defn binary_exp_recursive [binary_exp_recursive_base binary_exp_recursive_exponent]
  (binding [binary_exp_recursive_half nil] (try (do (when (< binary_exp_recursive_exponent 0) (throw (Exception. "exponent must be non-negative"))) (when (= binary_exp_recursive_exponent 0) (throw (ex-info "return" {:v 1.0}))) (when (= (mod binary_exp_recursive_exponent 2) 1) (throw (ex-info "return" {:v (* (binary_exp_recursive binary_exp_recursive_base (- binary_exp_recursive_exponent 1)) binary_exp_recursive_base)}))) (set! binary_exp_recursive_half (binary_exp_recursive binary_exp_recursive_base (/ binary_exp_recursive_exponent 2))) (throw (ex-info "return" {:v (* binary_exp_recursive_half binary_exp_recursive_half)}))) (catch clojure.lang.ExceptionInfo e (if (= (ex-message e) "return") (get (ex-data e) :v) (throw e))))))

(defn binary_exp_iterative [binary_exp_iterative_base binary_exp_iterative_exponent]
  (binding [binary_exp_iterative_b nil binary_exp_iterative_e nil binary_exp_iterative_result nil] (try (do (when (< binary_exp_iterative_exponent 0) (throw (Exception. "exponent must be non-negative"))) (set! binary_exp_iterative_result 1.0) (set! binary_exp_iterative_b binary_exp_iterative_base) (set! binary_exp_iterative_e binary_exp_iterative_exponent) (while (> binary_exp_iterative_e 0) (do (when (= (mod binary_exp_iterative_e 2) 1) (set! binary_exp_iterative_result (* binary_exp_iterative_result binary_exp_iterative_b))) (set! binary_exp_iterative_b (* binary_exp_iterative_b binary_exp_iterative_b)) (set! binary_exp_iterative_e (/ binary_exp_iterative_e 2)))) (throw (ex-info "return" {:v binary_exp_iterative_result}))) (catch clojure.lang.ExceptionInfo e (if (= (ex-message e) "return") (get (ex-data e) :v) (throw e))))))

(defn binary_exp_mod_recursive [binary_exp_mod_recursive_base binary_exp_mod_recursive_exponent binary_exp_mod_recursive_modulus]
  (binding [binary_exp_mod_recursive_r nil] (try (do (when (< binary_exp_mod_recursive_exponent 0) (throw (Exception. "exponent must be non-negative"))) (when (<= binary_exp_mod_recursive_modulus 0) (throw (Exception. "modulus must be positive"))) (when (= binary_exp_mod_recursive_exponent 0) (throw (ex-info "return" {:v (mod 1 binary_exp_mod_recursive_modulus)}))) (when (= (mod binary_exp_mod_recursive_exponent 2) 1) (throw (ex-info "return" {:v (mod (* (binary_exp_mod_recursive binary_exp_mod_recursive_base (- binary_exp_mod_recursive_exponent 1) binary_exp_mod_recursive_modulus) (mod binary_exp_mod_recursive_base binary_exp_mod_recursive_modulus)) binary_exp_mod_recursive_modulus)}))) (set! binary_exp_mod_recursive_r (binary_exp_mod_recursive binary_exp_mod_recursive_base (/ binary_exp_mod_recursive_exponent 2) binary_exp_mod_recursive_modulus)) (throw (ex-info "return" {:v (mod (* binary_exp_mod_recursive_r binary_exp_mod_recursive_r) binary_exp_mod_recursive_modulus)}))) (catch clojure.lang.ExceptionInfo e (if (= (ex-message e) "return") (get (ex-data e) :v) (throw e))))))

(defn binary_exp_mod_iterative [binary_exp_mod_iterative_base binary_exp_mod_iterative_exponent binary_exp_mod_iterative_modulus]
  (binding [binary_exp_mod_iterative_b nil binary_exp_mod_iterative_e nil binary_exp_mod_iterative_result nil] (try (do (when (< binary_exp_mod_iterative_exponent 0) (throw (Exception. "exponent must be non-negative"))) (when (<= binary_exp_mod_iterative_modulus 0) (throw (Exception. "modulus must be positive"))) (set! binary_exp_mod_iterative_result (mod 1 binary_exp_mod_iterative_modulus)) (set! binary_exp_mod_iterative_b (mod binary_exp_mod_iterative_base binary_exp_mod_iterative_modulus)) (set! binary_exp_mod_iterative_e binary_exp_mod_iterative_exponent) (while (> binary_exp_mod_iterative_e 0) (do (when (= (mod binary_exp_mod_iterative_e 2) 1) (set! binary_exp_mod_iterative_result (mod (* binary_exp_mod_iterative_result binary_exp_mod_iterative_b) binary_exp_mod_iterative_modulus))) (set! binary_exp_mod_iterative_b (mod (* binary_exp_mod_iterative_b binary_exp_mod_iterative_b) binary_exp_mod_iterative_modulus)) (set! binary_exp_mod_iterative_e (/ binary_exp_mod_iterative_e 2)))) (throw (ex-info "return" {:v binary_exp_mod_iterative_result}))) (catch clojure.lang.ExceptionInfo e (if (= (ex-message e) "return") (get (ex-data e) :v) (throw e))))))

(defn -main []
  (let [rt (Runtime/getRuntime)
    start-mem (- (.totalMemory rt) (.freeMemory rt))
    start (System/nanoTime)]
      (println (binary_exp_recursive 3.0 5))
      (println (binary_exp_iterative 1.5 4))
      (println (binary_exp_mod_recursive 3 4 5))
      (println (binary_exp_mod_iterative 11 13 7))
      (System/gc)
      (let [end (System/nanoTime)
        end-mem (- (.totalMemory rt) (.freeMemory rt))
        duration-us (quot (- end start) 1000)
        memory-bytes (Math/abs ^long (- end-mem start-mem))]
        (println (str "{\n  \"duration_us\": " duration-us ",\n  \"memory_bytes\": " memory-bytes ",\n  \"name\": \"main\"\n}"))
      )
    ))

(-main)

File: tests/algorithms/x/Pascal/maths/binary_exponentiation.pas
{$mode objfpc}{$modeswitch nestedprocvars}
program Main;
uses SysUtils;
var _nowSeed: int64 = 0;
var _nowSeeded: boolean = false;
procedure init_now();
var s: string; v: int64;
begin
  s := GetEnvironmentVariable('MOCHI_NOW_SEED');
  if s <> '' then begin
    Val(s, v);
    _nowSeed := v;
    _nowSeeded := true;
  end;
end;
function _now(): integer;
begin
  if _nowSeeded then begin
    _nowSeed := (_nowSeed * 1664525 + 1013904223) mod 2147483647;
    _now := _nowSeed;
  end else begin
    _now := Integer(GetTickCount64()*1000);
  end;
end;
function _bench_now(): int64;
begin
  _bench_now := GetTickCount64()*1000;
end;
function _mem(): int64;
var h: TFPCHeapStatus;
begin
  h := GetFPCHeapStatus;
  _mem := h.CurrHeapUsed;
end;
procedure panic(msg: string);
begin
  writeln(msg);
  halt(1);
end;
procedure error(msg: string);
begin
  panic(msg);
end;
function _to_float(x: integer): real;
begin
  _to_float := x;
end;
function to_float(x: integer): real;
begin
  to_float := _to_float(x);
end;
procedure json(xs: array of real);
var i: integer;
begin
  write('[');
  for i := 0 to High(xs) do begin
    write(xs[i]);
    if i < High(xs) then write(', ');
  end;
  writeln(']');
end;
var
  bench_start_0: integer;
  bench_dur_0: integer;
  bench_mem_0: int64;
  bench_memdiff_0: int64;
  modulus: integer;
  base: integer;
  exponent: integer;
function binary_exp_recursive(base: real; exponent: integer): real; forward;
function binary_exp_iterative(base: real; exponent: integer): real; forward;
function binary_exp_mod_recursive(base: integer; exponent: integer; modulus: integer): integer; forward;
function binary_exp_mod_iterative(base: integer; exponent: integer; modulus: integer): integer; forward;
function binary_exp_recursive(base: real; exponent: integer): real;
var
  binary_exp_recursive_half: real;
begin
  if exponent < 0 then begin
  panic('exponent must be non-negative');
end;
  if exponent = 0 then begin
  exit(1);
end;
  if (exponent mod 2) = 1 then begin
  exit(binary_exp_recursive(base, exponent - 1) * base);
end;
  binary_exp_recursive_half := binary_exp_recursive(base, exponent div 2);
  exit(binary_exp_recursive_half * binary_exp_recursive_half);
end;
function binary_exp_iterative(base: real; exponent: integer): real;
var
  binary_exp_iterative_result_: real;
  binary_exp_iterative_b: real;
  binary_exp_iterative_e: integer;
begin
  if exponent < 0 then begin
  panic('exponent must be non-negative');
end;
  binary_exp_iterative_result_ := 1;
  binary_exp_iterative_b := base;
  binary_exp_iterative_e := exponent;
  while binary_exp_iterative_e > 0 do begin
  if (binary_exp_iterative_e mod 2) = 1 then begin
  binary_exp_iterative_result_ := binary_exp_iterative_result_ * binary_exp_iterative_b;
end;
  binary_exp_iterative_b := binary_exp_iterative_b * binary_exp_iterative_b;
  binary_exp_iterative_e := binary_exp_iterative_e div 2;
end;
  exit(binary_exp_iterative_result_);
end;
function binary_exp_mod_recursive(base: integer; exponent: integer; modulus: integer): integer;
var
  binary_exp_mod_recursive_r: integer;
begin
  if exponent < 0 then begin
  panic('exponent must be non-negative');
end;
  if modulus <= 0 then begin
  panic('modulus must be positive');
end;
  if exponent = 0 then begin
  exit(1 mod modulus);
end;
  if (exponent mod 2) = 1 then begin
  exit((binary_exp_mod_recursive(base, exponent - 1, modulus) * (base mod modulus)) mod modulus);
end;
  binary_exp_mod_recursive_r := binary_exp_mod_recursive(base, exponent div 2, modulus);
  exit((binary_exp_mod_recursive_r * binary_exp_mod_recursive_r) mod modulus);
end;
function binary_exp_mod_iterative(base: integer; exponent: integer; modulus: integer): integer;
var
  binary_exp_mod_iterative_result_: integer;
  binary_exp_mod_iterative_b: integer;
  binary_exp_mod_iterative_e: integer;
begin
  if exponent < 0 then begin
  panic('exponent must be non-negative');
end;
  if modulus <= 0 then begin
  panic('modulus must be positive');
end;
  binary_exp_mod_iterative_result_ := 1 mod modulus;
  binary_exp_mod_iterative_b := base mod modulus;
  binary_exp_mod_iterative_e := exponent;
  while binary_exp_mod_iterative_e > 0 do begin
  if (binary_exp_mod_iterative_e mod 2) = 1 then begin
  binary_exp_mod_iterative_result_ := (binary_exp_mod_iterative_result_ * binary_exp_mod_iterative_b) mod modulus;
end;
  binary_exp_mod_iterative_b := (binary_exp_mod_iterative_b * binary_exp_mod_iterative_b) mod modulus;
  binary_exp_mod_iterative_e := binary_exp_mod_iterative_e div 2;
end;
  exit(binary_exp_mod_iterative_result_);
end;
begin
  init_now();
  bench_mem_0 := _mem();
  bench_start_0 := _bench_now();
  writeln(binary_exp_recursive(3, 5));
  writeln(binary_exp_iterative(1.5, 4));
  writeln(binary_exp_mod_recursive(3, 4, 5));
  writeln(binary_exp_mod_iterative(11, 13, 7));
  bench_memdiff_0 := _mem() - bench_mem_0;
  bench_dur_0 := (_bench_now() - bench_start_0) div 1000;
  writeln('{');
  writeln(('  "duration_us": ' + IntToStr(bench_dur_0)) + ',');
  writeln(('  "memory_bytes": ' + IntToStr(bench_memdiff_0)) + ',');
  writeln(('  "name": "' + 'main') + '"');
  writeln('}');
end.
